Assets/Mirror/Runtime/SyncDictionary.cs

```diff
@@ -0,0 +1,297 @@
+        public event SyncDictionaryChanged Callback;
+
+        public enum Operation : byte
+        {
+            OP_ADD,
+            OP_CLEAR,
+            OP_REMOVE,
+            OP_SET,
+            OP_DIRTY
+        }
+
+        struct Change
+        {
+            internal Operation operation;
+            internal B key;
+            internal T item;
+        }
+
+        readonly List<Change> Changes = new List<Change>();
+        // how many changes we need to ignore
+        // this is needed because when we initialize the list,
+        // we might later receive changes that have already been applied
+        // so we need to skip them
+        int changesAhead = 0;
+
+        protected virtual void SerializeKey(NetworkWriter writer, B item) {}
+        protected virtual void SerializeItem(NetworkWriter writer, T item) {}
+        protected virtual B DeserializeKey(NetworkReader reader) => default;
+        protected virtual T DeserializeItem(NetworkReader reader) => default;
+
+        public bool IsDirty => Changes.Count > 0;
+
+        public ICollection<B> Keys => m_Objects.Keys;
+
+        public ICollection<T> Values => m_Objects.Values;
+
+        // throw away all the changes
+        // this should be called after a successfull sync
+        public void Flush() => Changes.Clear();
+
+        public SyncDictionary()
+        {
+            m_Objects = new Dictionary<B, T>();
+        }
+
+        public SyncDictionary(IEqualityComparer<B> eq)
+        {
+            m_Objects = new Dictionary<B, T>(eq);
+        }
+
+        void AddOperation(Operation op, B key, T item)
+        {
+            if (IsReadOnly)
+            {
+                throw new System.InvalidOperationException("Synclists can only be modified at the server");
+            }
+
+            Change change = new Change
+            {
+                operation = op,
+                key = key,
+                item = item
+            };
+
+            Changes.Add(change);
+
+            Callback?.Invoke(op, key, item);
+        }
+
+        public void OnSerializeAll(NetworkWriter writer)
+        {
+            // if init,  write the full list content
+            writer.WritePackedUInt32((uint)m_Objects.Count);
+
+            foreach (KeyValuePair<B, T> syncItem in m_Objects)
+            {
+                SerializeKey(writer, syncItem.Key);
+                SerializeItem(writer, syncItem.Value);
+            }
+
+            // all changes have been applied already
+            // thus the client will need to skip all the pending changes
+            // or they would be applied again.
+            // So we write how many changes are pending
+            writer.WritePackedUInt32((uint)Changes.Count);
+        }
+
+        public void OnSerializeDelta(NetworkWriter writer)
+        {
+            // write all the queued up changes
+            writer.WritePackedUInt32((uint)Changes.Count);
+
+            for (int i = 0; i < Changes.Count; i++)
```

Comment:
++i instead of i++ is more obvious imho. that's what it's supposed to do, right?

Reply:
Doing ++ after is more common practice

Reply:
for (int i = 0; i < changes.Count; (i += 1)-1) would be way more strange than
for (int i = 0; i < changes.Count; i += 1) if you ask me, but will leave that one up to katori

Reply:
i++ is way more common.

Reply:
I respect where you're coming from vis, but as others have said, i++ is the standard in C# (and many languages), and the compiler difference is possibly nonexistent. Leaving for now.